src/relayclient/Enveloping.ts

```diff
@@ -0,0 +1,236 @@
+
+import { DeployTransactionRequest, RelayMetadata, RelayTransactionRequest } from './types/RelayTransactionRequest'
+import HttpWrapper from './HttpWrapper'
+import { constants } from '../common/Constants'
+import { HttpProvider } from 'web3-core'
+import { RelayingAttempt } from './RelayClient'
+
+import { getDependencies, EnvelopingConfig, EnvelopingDependencies } from './Configurator'
+import { Address, IntString } from './types/Aliases'
+
+import { PrefixedHexString, Transaction } from 'ethereumjs-tx'
+
+import TypedRequestData, { getDomainSeparatorHash, TypedDeployRequestData } from '../common/EIP712/TypedRequestData'
+
+const zeroAddr = '0x0000000000000000000000000000000000000000'
+
+export interface SignatureProvider {
+  sign: (dataToSign: TypedRequestData) => PrefixedHexString
+  verifySign: (signature: PrefixedHexString, dataToSign: TypedRequestData, request: RelayRequest|DeployRequest) => boolean
+}
+
+export class Enveloping {
+  config: EnvelopingConfig
+  relayWorkerAddress: Address
+  dependencies: EnvelopingDependencies
+  private initialized: boolean
+
+  constructor (_config: EnvelopingConfig, _web3: Web3, _relayWorkerAddress: Address) {
+    this.config = _config
+    this.initialized = false
+    this.dependencies = getDependencies(this.config, _web3.currentProvider as HttpProvider)
+    this.relayWorkerAddress = _relayWorkerAddress
+  }
+
+  async _init (): Promise<void> {
+    if (!this.initialized) {
+      await this.dependencies.contractInteractor.init()
+      this.initialized = true
+    } else {
+      throw new Error('_init was already called')
+    }
+  }
+
+  /**
+    * creates a deploy request
+    * @param from - sender's wallet address (EOA)
+    * @param gasLimit - gas limit of the relayed transaction
+    * @param tokenContract - the token the user will use to pay to the Worker for deploying the SmartWallet
+    * @param tokenAmount - the token amount the user will pay for the deployment, zero if the deploy is subsidized
+    * @param tokenGas - gas limit of the token payment
+    * @param  gasPrice - optional: if not set the gasPrice is calculated internally through a call to estimateGas
+    * @param  index - optional: allows the user to create multiple SmartWallets
+    * @param  recoverer optional: This SmartWallet instance won't have recovery support
+    * @return a deploy request structure.
+    */
+  async createDeployRequest (from: Address, gasLimit: IntString, tokenContract: Address, tokenAmount: IntString, tokenGas: IntString, gasPrice?: IntString, index? : IntString, recoverer? : IntString): Promise<DeployRequest> {
+    const deployRequest: DeployRequest = {
+      request: {
+        relayHub: this.config.relayHubAddress,
+        from: from,
+        to: zeroAddr,
+        value: '0',
+        gas: gasLimit, // overhead (cte) + fee + (estimateDeploy * 1.1)
+        nonce: (await this.getFactoryNonce(this.config.smartWalletFactoryAddress, from)).toString(),
```

Comment:
if factory is fetched from config then the deployRequest is only for that type of smart wallet